pkg/cli/admin/release/extract_tools.go

```diff
@@ -1281,8 +1369,66 @@ func findClusterIncludeConfig(ctx context.Context, restConfig *rest.Config) (man
 	return config, nil
 }
 
-func newIncluder(config manifestInclusionConfiguration) includer {
-	return func(m *manifest.Manifest) error {
-		return m.Include(config.ExcludeIdentifier, config.RequiredFeatureSet, config.Profile, config.Capabilities, config.Overrides)
+// getCurrentPayloadManifests gets the manifests that are reconciled on the cluster
+func getCurrentPayloadManifests(image string, inclusionConfig manifestInclusionConfiguration) []manifest.Manifest {
+	// TODO use a new opts to retrieve the manifests for the release image that is running on the cluster and
+	// use the inclusionConfig to determine if it is included on the cluster
+	return nil
+}
+
+// GetImplicitlyEnabledCapabilities returns a set of capabilities that are implicitly enabled after a cluster update.
+// The arguments are two sets of manifests, manifest inclusion configuration, and
+// a set of capabilities that are implicitly enabled on the cluster, i.e., the capabilities
+// that are NOT specified in the cluster version but has to considered enabled on the cluster.
+// The manifest inclusion configuration is used to determine if a manifest should be included.
+// In other words, whether, or not the cluster version operator reconcile that manifest on the cluster.
+// The two sets of manifests are respectively from the release that is currently running on the cluster and
+// from the release that the cluster is updated to.
+// TODO lift this function to library-go
+func GetImplicitlyEnabledCapabilities(
+	updatePayloadManifests []manifest.Manifest,
+	currentPayloadManifests []manifest.Manifest,
+	manifestInclusionConfiguration manifestInclusionConfiguration,
+	currentImplicitlyEnabled sets.Set[configv1.ClusterVersionCapability],
+) sets.Set[configv1.ClusterVersionCapability] {
+
+	ret := sets.New[configv1.ClusterVersionCapability]().Union(currentImplicitlyEnabled)
+	for _, updateManifest := range updatePayloadManifests {
+		updateManErr := updateManifest.IncludeAllowUnknownCapabilities(
+			manifestInclusionConfiguration.ExcludeIdentifier,
+			manifestInclusionConfiguration.RequiredFeatureSet,
+			manifestInclusionConfiguration.Profile,
+			manifestInclusionConfiguration.Capabilities,
+			manifestInclusionConfiguration.Overrides,
+			true,
+		)
+		// update manifest is enabled, no need to check
```

Comment:
hmm, but with `IncludeAllowUnknownCapabilities`, aren't the ones that fail the ones we can ignore?  I'd guess they meant "not even with all these new caps enabled will this manifest get included".  The ones we'll implicitly enable caps for are the ones that fail to get included if we apply the current caps, but which do get included if we allow new caps, and which match `SameResourceID`.

Reply:
This function is refactored from this one in CVO: https://github.com/openshift/cluster-version-operator/blob/e90705bb5457b2b9d449b4377036c3de6617ebb6/pkg/payload/payload.go#L247-L252

My understanding is very weak on this function (although i have read it a couple of times already).

> hmm, but with IncludeAllowUnknownCapabilities, aren't the ones that fail the ones we can ignore?

No. Passed ones (`updateManErr == nil` in the next line) are ignored. The manifests that are included on the cluster wont generate new implicitly enabled capabilities.